zebra/if_netlink.c

```diff
@@ -350,11 +350,38 @@ static void netlink_vrf_change(struct nlmsghdr *h, struct rtattr *tb,
 			}
 		}
 
-		/*
-		 * vrf_get is implied creation if it does not exist
+		/*Treat VRF add for existing vrf as update
+		 * Update VRF ID and also update in VRF ID table
 		 */
-		vrf = vrf_get((vrf_id_t)ifi->ifi_index,
-			      name); // It would create vrf
+		if (name)
+			vrf = vrf_lookup_by_name(name);
+		if (vrf && (vrf_id_t)ifi->ifi_index != VRF_UNKNOWN
+		    && vrf->vrf_id != VRF_UNKNOWN
+		    && vrf->vrf_id != (vrf_id_t)ifi->ifi_index) {
+			zlog_info(
+				"Vrf Update event: %s old id: %u table:%u new id: %u table: %u",
+				name, vrf->vrf_id,
+				((struct zebra_vrf *)vrf->info)->table_id,
+				(vrf_id_t)ifi->ifi_index, nl_table_id);
+
+			/*Disable the vrf to simulate implicit delete
+			 * so that all stale routes are deleted
+			 * This vrf will be enabled down the line*/
+			vrf_disable(vrf);
```

Comment:
Yes, agree with Stephen - we need to separate the netlink-level message handling from the higher-level zebra processing of the event. In the future, these netlink events may not even be processed in the main zebra pthread.

Reply:
I have done the changes.